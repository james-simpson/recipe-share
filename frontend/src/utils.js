export function getRecipeDurationLabel (timeInMins) {

	let hours = Math.floor(timeInMins / 60);
  let minutes = timeInMins % 60;
  if (hours > 0 && minutes > 0) {
    return hours + 'h ' + minutes + 'm';
  } else if (hours > 0) {
    return hours + 'h'
  }

  return minutes + 'm';
}
