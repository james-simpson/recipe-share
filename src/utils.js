export function getRecipeDurationLabel (timeInMins) {
  let hours = Math.floor(timeInMins / 60)
  let minutes = timeInMins % 60
  if (hours > 0 && minutes > 0) {
    return hours + 'h ' + minutes + 'm'
  } else if (hours > 0) {
    return hours + 'h'
  }

  return minutes + 'm'
}

export function getDifficultyLabel (difficultyIndex) {
  const difficultyLevels = ['Easy', 'Easy/medium', 'Medium', 'Medium/Advanced', 'Advanced']
  return difficultyLevels[difficultyIndex]
}
