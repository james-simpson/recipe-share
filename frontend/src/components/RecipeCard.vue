<template>
  <v-card :color="color" class="white--text" :to="to" hover="hover">
    <v-container fluid grid-list-lg>
      <v-layout row>
        <v-flex xs7>
          <div>
            <div class="title">{{ title }}</div>
            <div class="pb-2">{{ author }}</div>
            <v-chip
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              {{ timeLabel }}
            </v-chip>
            <v-chip
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              {{ difficultyLevel }}
            </v-chip>
            <v-chip
              v-if="vegetarian"
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              Vegetarian
            </v-chip>
          </div>
        </v-flex>
        <v-flex xs5>
          <v-card-media
            :src="imagePath"
            height="120px"
            contain
            center
          ></v-card-media>
        </v-flex>
      </v-layout>
    </v-container>
  </v-card>
</template>

<script>
  import { getRecipeDurationLabel } from '../utils.js';

  export default {
    name: 'RecipeCard',
    props: [
      'title',
      'author',
      'image',
      'time',
      'difficulty',
      'difficultyLevels',
      'vegetarian',
      'vegan',
      'sweet',
      'to'
    ],
    computed: {
      color: function() {
        if (this.sweet === true) {
          return 'pink lighten-2'
        }

        if (this.vegetarian === true) {
          return 'green';
        }

        return 'cyan darken-1';
      },
      imagePath: function() {
        return '/static/' + this.image;
      },
      difficultyLevel: function() {
        return this.difficultyLevels[this.difficulty];
      },
      timeLabel () {
        return getRecipeDurationLabel(this.time);
      }
    }
  }
</script>

<style >
.loading-icon {
  text-align: center;
}      
</style>
