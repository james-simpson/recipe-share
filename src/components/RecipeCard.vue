<template>
  <v-card :color="color" class="white--text" :to="to" hover="hover">
    <v-container fluid grid-list-lg>
      <v-layout row>
        <v-flex xs7>
          <div>
            <div class="title">{{ title }}</div>
            <div class="pb-2">{{ author }}</div>
            <v-chip
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              {{ timeLabel }}
            </v-chip>
            <v-chip
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              {{ difficultyLevel }}
            </v-chip>
            <v-chip
              v-if="vegetarian && !vegan"
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              Vegetarian
            </v-chip>
            <v-chip
              v-if="vegan"
              color="grey darken-3"
              text-color="white"
              small
              disabled
            >
              Vegan
            </v-chip>
          </div>
        </v-flex>
        <v-flex xs5>
          <v-img
            :src="image"
            height="120px"
            contain
            center
          ></v-img>
        </v-flex>
      </v-layout>
    </v-container>
    <v-btn
      @click="toggleShoppingListRecipeIds($event, id)"
      fab
      top
      right
      dark
      :color="listButtonColor"
      class="shopping-list-button"
    >
      <v-icon v-if="inShoppingList" dark>done</v-icon>
      <v-icon v-else>add_shopping_cart</v-icon>
    </v-btn>
  </v-card>
</template>

<script>
import { getRecipeDurationLabel, getDifficultyLabel } from '../utils.js'

export default {
  name: 'RecipeCard',
  props: [
    'id',
    'title',
    'author',
    'image',
    'time',
    'difficulty',
    'difficultyLevels',
    'vegetarian',
    'vegan',
    'sweet',
    'to',
    'shoppingListIds',
    'toggleShoppingListRecipeIds'
  ],
  computed: {
    color () {
      if (this.sweet === true) {
        return 'pink lighten-2'
      }

      if (this.vegetarian === true) {
        return 'green'
      }

      return 'cyan darken-1'
    },
    difficultyLevel () {
      return getDifficultyLabel(this.difficulty)
    },
    timeLabel () {
      return getRecipeDurationLabel(this.time)
    },
    inShoppingList () {
      return this.shoppingListIds.includes(this.id)
    },
    listButtonColor () {
      return this.inShoppingList ? 'green' : 'primary'
    }
  }
}
</script>

<style >
.loading-icon {
  text-align: center;
}
.shopping-list-button {
  position: absolute !important;
  top: 5px !important;
  right: 5px !important;
}
</style>
