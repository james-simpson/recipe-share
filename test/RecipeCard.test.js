import { shallow, mount } from 'vue-test-utils'
import RecipeCard from '../src/components/RecipeCard'
import { getRecipeDurationLabel, getDifficultyLabel } from '../src/utils.js'

jest.mock('../src/utils.js', () => ({
  getDifficultyLabel: jest.fn(),
  getRecipeDurationLabel: jest.fn(() => 'duration label')
}))

const createCmp = propsData => shallow(RecipeCard, { propsData })
const createMountedCmp = propsData => mount(RecipeCard, { propsData })

describe('RecipeCard', () => {
  let cmp
  const defaultProps = {
    title: 'Egg custard tarts',
    author: 'Gregg Wallace',
    time: 60,
    difficulty: 3,
    vegetarian: false,
    vegan: false,
  }

  cmp = createCmp(defaultProps);

  beforeEach(() => {
    jest.resetModules()
    jest.resetAllMocks()
  })

  describe('HTML', () => {
    const mountedCmp = createMountedCmp(defaultProps)
    it('is as expected', () => {
      expect(mountedCmp.vm.$el).toMatchSnapshot()
    })

    it('is as expected when recipe is vegetarian', () => {
      const mountedCmp = createMountedCmp(defaultProps)
      cmp = createCmp({ vegetarian: true })
      expect(mountedCmp.vm.$el).toMatchSnapshot()
    })

    it('is as expected when recipe is vegan', () => {
      cmp = createCmp({ vegetarian: true, vegan: true })
      expect(cmp.vm.$el).toMatchSnapshot()
    })
  })

  describe('Props', () => {
    cmp = createCmp({ })

    describe('title', () => {
      const title = cmp.vm.$options.props.title

      it('is required', () => {
        expect(title.required).toEqual(true)
      })

      it('is a string', () => {
        expect(title.type).toEqual(String)
      })
    })

    describe('author', () => {
      const author = cmp.vm.$options.props.author

      it('is required', () => {
        expect(author.required).toEqual(true)
      })

      it('is a string', () => {
        expect(author.type).toEqual(String)
      })
    })

    describe('image', () => {
      const image = cmp.vm.$options.props.image

      it('is not required', () => {
        expect(image.required).toBe(false)
      })

      it('defaults to undefined', () => {
        expect(cmp.props().image).toBe(undefined)
      })

      it('is a string', () => {
        expect(image.type).toEqual(String)
      })
    })

    describe('time', () => {
      const time = cmp.vm.$options.props.time

      it('is required', () => {
        expect(time.required).toEqual(true)
      })

      it('is a number', () => {
        expect(time.type).toEqual(Number)
      })
    })

    describe('difficulty', () => {
      const difficulty = cmp.vm.$options.props.difficulty

      it('is required', () => {
        expect(difficulty.required).toEqual(true)
      })

      it('is a number', () => {
        expect(difficulty.type).toEqual(Number)
      })
    })

    describe('vegetarian', () => {
      const vegetarian = cmp.vm.$options.props.vegetarian

      it('is required', () => {
        expect(vegetarian.required).toEqual(true)
      })

      it('is a boolean', () => {
        expect(vegetarian.type).toEqual(Boolean)
      })
    })

    describe('vegan', () => {
      const vegan = cmp.vm.$options.props.vegan

      it('is required', () => {
        expect(vegan.required).toEqual(true)
      })

      it('is a boolean', () => {
        expect(vegan.type).toEqual(Boolean)
      })
    })

    describe('sweet', () => {
      const sweet = cmp.vm.$options.props.sweet

      it('is required', () => {
        expect(sweet.required).toEqual(true)
      })

      it('is a boolean', () => {
        expect(sweet.type).toEqual(Boolean)
      })
    })

    describe('to', () => {
      const toLink = cmp.vm.$options.props.to

      it('is required', () => {
        expect(toLink.required).toEqual(true)
      })

      it('is a string', () => {
        expect(toLink.type).toEqual(String)
      })
    })
  })

  describe('Computed properties', () => {
    describe('color', () => {
      it('is correct if recipe is not vegetarian and not sweet', () => {
        cmp = createCmp({ vegetarian: false })
        expect(cmp.vm.color).toBe('cyan darken-1')
      })

      it('is correct if recipe is vegetarian and not sweet', () => {
        cmp = createCmp({ vegetarian: true })
        expect(cmp.vm.color).toBe('green')
      })

      it('is correct if recipe is sweet and vegetarian', () => {
        cmp = createCmp({ sweet: true, vegetarian: true })
        expect(cmp.vm.color).toBe('pink lighten-2')
      })

      it('is correct if recipe is sweet and not vegetarian', () => {
        cmp = createCmp({ sweet: true, vegetarian: false })
        expect(cmp.vm.color).toBe('pink lighten-2')
      })
    })

    describe('difficultyLabel', () => {
      it('calls getDifficultyLabel and returns the result', () => {
        getDifficultyLabel.mockReturnValue('difficulty label')
        cmp = createCmp(defaultProps)
        expect(getDifficultyLabel).toBeCalledWith(3)
        expect(cmp.vm.difficultyLabel).toBe('difficulty label')
      })
    })

    describe('timeLabel', () => {
      it('calls getRecipeDurationLabel and returns the result', () => {
        getRecipeDurationLabel.mockReturnValue('duration label')
        cmp = createCmp(defaultProps)
        expect(getRecipeDurationLabel).toBeCalledWith(60)
        expect(cmp.vm.timeLabel).toBe('duration label')
      })
    })
  })
})
