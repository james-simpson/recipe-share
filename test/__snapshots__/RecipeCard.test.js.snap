// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard HTML is as expected 1`] = `
<v-card
  class="white--text"
  color="cyan darken-1"
  hover="hover"
>
  <v-container
    fluid=""
    grid-list-lg=""
  >
    <v-layout
      row=""
    >
      <v-flex
        xs7=""
      >
        <div>
          <div
            class="title"
          >
            Egg custard tarts
          </div>
           
          <div
            class="pb-2"
          >
            Gregg Wallace
          </div>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            duration label
          
          </v-chip>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            
          
          </v-chip>
           
          <!---->
           
          <!---->
        </div>
      </v-flex>
       
      <v-flex
        xs5=""
      >
        <v-card-media
          center=""
          contain=""
          height="120px"
        />
      </v-flex>
    </v-layout>
  </v-container>
</v-card>
`;

exports[`RecipeCard HTML is as expected when recipe is vegan 1`] = `
<v-card
  class="white--text"
  color="green"
  hover="hover"
>
  <v-container
    fluid=""
    grid-list-lg=""
  >
    <v-layout
      row=""
    >
      <v-flex
        xs7=""
      >
        <div>
          <div
            class="title"
          >
            
          </div>
           
          <div
            class="pb-2"
          >
            
          </div>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            
          
          </v-chip>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            
          
          </v-chip>
           
          <!---->
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            Vegan
          
          </v-chip>
        </div>
      </v-flex>
       
      <v-flex
        xs5=""
      >
        <v-card-media
          center=""
          contain=""
          height="120px"
        />
      </v-flex>
    </v-layout>
  </v-container>
</v-card>
`;

exports[`RecipeCard HTML is as expected when recipe is vegetarian 1`] = `
<v-card
  class="white--text"
  color="cyan darken-1"
  hover="hover"
>
  <v-container
    fluid=""
    grid-list-lg=""
  >
    <v-layout
      row=""
    >
      <v-flex
        xs7=""
      >
        <div>
          <div
            class="title"
          >
            Egg custard tarts
          </div>
           
          <div
            class="pb-2"
          >
            Gregg Wallace
          </div>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            
          
          </v-chip>
           
          <v-chip
            color="grey darken-3"
            disabled="disabled"
            small=""
            text-color="white"
          >
            
            
          
          </v-chip>
           
          <!---->
           
          <!---->
        </div>
      </v-flex>
       
      <v-flex
        xs5=""
      >
        <v-card-media
          center=""
          contain=""
          height="120px"
        />
      </v-flex>
    </v-layout>
  </v-container>
</v-card>
`;
